### PR TITLE
Feature: add warehouse grammar and GUC variable

### PR DESCRIPTION
**Pull request by roseduan**
### Change logs

Since we are going to support storage/compute/catalog separation, we
need to add a warehouse logic to the system. Then we can use the
warehouse clause to specify the warehouse for a SQL query. A warehouse
is a group of segments, and we can create/alter/delete a warehouse.

This commit does the following things:

1. add the grammar for the warehouse clause.
2. add the GUC variable for the current warehouse.
3. add some fileter conditions for gp_segment_configuration table, since we
   need to filter out the segments that are not in the current warehouse.

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to @cloudberrydb/dev team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/backend/parser/gram.y [hunk 1]
```diff
@@ -294,7 +294,7 @@ static void check_expressions_in_partition_key(PartitionSpec *spec, core_yyscan_
 		CreateFdwStmt CreateForeignServerStmt CreateForeignTableStmt
 		CreateAssertionStmt CreateTransformStmt CreateTrigStmt CreateEventTrigStmt
 		CreateUserStmt CreateUserMappingStmt CreateRoleStmt CreatePolicyStmt
-		CreatedbStmt DeclareCursorStmt DefineStmt DeleteStmt DiscardStmt DoStmt
+		CreatedbStmt CreateWarehouseStmt DeclareCursorStmt DefineStmt DeleteStmt DiscardStmt DoStmt
```
Comment (my-ship-it) [inline src/backend/parser/gram.y:297]: Please also add create_warehouse.sgml?

File: src/backend/cdb/cdbutil.c [hunk 2]
```diff
@@ -1507,12 +1552,20 @@ dbid_get_dbinfo(int16 dbid)
...
 
 	/* SELECT * FROM gp_segment_configuration WHERE dbid = :1 */
-	ScanKeyInit(&scankey,
+	ScanKeyInit(&scankey[0],
 				Anum_gp_segment_configuration_dbid,
 				BTEqualStrategyNumber, F_INT2EQ,
 				Int16GetDatum(dbid));
-	scan = systable_beginscan(rel, GpSegmentConfigDbidIndexId, true,
-							  NULL, 1, &scankey);
+	if (dbid != 1)
```
Comment (my-ship-it) [inline src/backend/cdb/cdbutil.c:1559]: Can we use ScalarArrayOpExpr here?

File: src/backend/cdb/cdbutil.c [hunk 3]
```diff
@@ -952,6 +957,45 @@ cdbcomponent_activeQEsExist(void)
...
+	sscan = systable_beginscan(rel, InvalidOid, false, NULL, 0, NULL);
+	while ((tuple = systable_getnext(sscan)) != NULL)
+	{
+		char *warehouse_name = NULL;
+		attr = heap_getattr(tuple, Anum_gp_segment_configuration_warehouse_name,
+							RelationGetDescr(rel), &isNull);
+		if (!isNull)
+			warehouse_name = TextDatumGetCString(attr);
+
+		if (warehouse_name && strcmp(warehouse_name, new_name) == 0)
```
Comment (my-ship-it) [inline src/backend/cdb/cdbutil.c:984]: Is it possible drop warehouse?

Comment (my-ship-it) [reply src/backend/cdb/cdbutil.c:984]: We need to process if warehouse is used by another session, we cannot drop that warehouse by another session.

File: src/backend/utils/misc/guc_gp.c [hunk 4]
```diff
@@ -5079,6 +5093,22 @@ check_max_running_tasks(int *newval, void **extra, GucSource source)
...
+	return cdb_checkWarehouseName(*newval);
+#else
+	return true;
+#endif
+}
+
+void assign_current_warehouse(const char *newval, void *extra)
+{
+	/* clear cache after warehouse changed */
+	cdbcomponent_destroyCdbComponents();
```
Comment (my-ship-it) [inline src/backend/utils/misc/guc_gp.c:5109]: When switch to a new warehouse, we need to lock the warehouse entry by read lock?

File: src/include/nodes/parsenodes.h [hunk 5]
```diff
@@ -4260,4 +4260,15 @@ typedef struct RetrieveStmt
 	bool		is_all;
 } RetrieveStmt;
 
+/* ----------------------
+ *		Warehouse Statement
+ * ----------------------
+ */
+typedef struct CreateWarehouseStmt
```
Comment (HuSen8891) [inline src/include/nodes/parsenodes.h:4267]: Add _copyCreateWarehouseStmt、_equalCreateWarehouseStmt and _outCreateWarehouseStmt(if needed) for CreateWarehouseStmt?

File: src/backend/utils/gp/segadmin.c [hunk 6]
```diff
@@ -194,13 +194,27 @@ remove_segment_config(int16 dbid)
...
+
+		attr = heap_getattr(tuple, Anum_gp_segment_configuration_warehouse_name,
+							RelationGetDescr(rel), &isNull);
+		if (!isNull)
+			warehouse_name = TextDatumGetCString(attr);
+
+		attr = heap_getattr(tuple, Anum_gp_segment_configuration_content,
+							RelationGetDescr(rel), &isNull);
+		Assert(!isNull);
+		if (DatumGetInt16(attr) == MASTER_CONTENT_ID || strcmp(warehouse_name, current_warehouse) == 0)
```
Comment (HuSen8891) [inline src/backend/utils/gp/segadmin.c:213]: why need this  (DatumGetInt16(attr) == MASTER_CONTENT_ID) condition？

File: src/backend/tcop/utility.c [hunk 7]
```diff
@@ -3729,6 +3729,10 @@ CreateCommandTag(Node *parsetree)
 			tag = CMDTAG_RETRIEVE;
 			break;
 
+		case T_CreateWarehouseStmt:
```
Comment (HuSen8891) [inline src/backend/tcop/utility.c:3732]: handle T_CreateWarehouseStmt in GetCommandLogLevel/ClassifyUtilityCommandAsReadOnly, also make check in other places.

File: src/include/nodes/parsenodes.h (hunk 5 above)
Comment (roseduan) [reply src/include/nodes/parsenodes.h:4267]: The `WarehouseStmt` will not be dispatched to Segments, I think it is not necessary.(Or we can add it when we need it)

File: src/backend/utils/gp/segadmin.c (hunk 6 above)
Comment (roseduan) [reply src/backend/utils/gp/segadmin.c:213]: warehouse_name of the master is null, when we query from the `gp_sgment_configuration`, this clause is like `where content == -1 or current_warehouse = warehouse_name`.

File: src/backend/catalog/gp_warehouse.c [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+#include "access/table.h"
+#include "catalog/catalog.h"
+#include "catalog/indexing.h"
+#include "catalog/gp_warehouse.h"
+
+#include "utils/builtins.h"
+#include "utils/rel.h"
+
+void
+GpWarehouseCreate(char *warehouse_name)
```
Comment (my-ship-it) [inline src/backend/catalog/gp_warehouse.c:14]: We don't need to implement GpWarehouseCreate in CBDB?

File: src/backend/cdb/cdbutil.c [hunk 9]
```diff
@@ -952,6 +964,75 @@ cdbcomponent_activeQEsExist(void)
...
+	Oid warehouseid = GetGpWarehouseOid(new_name, true);
+
+	if (!OidIsValid(warehouseid))
+		GUC_check_errmsg("warehouse %s does not exist", new_name);
+
+	return true;
+}
+
+void
+cdb_assignCurrentWarehouse(char *old_name, char *new_name)
```
Comment (my-ship-it) [inline src/backend/cdb/cdbutil.c:984]: We don't need to  place cdb_assignCurrentWarehouse in CBDB.

Comment (roseduan) [reply src/backend/cdb/cdbutil.c:986]: when we set the GUC value, we can not call this function from the extension.

File: src/backend/catalog/gp_warehouse.c (hunk 8 above)
Comment (roseduan) [reply src/backend/catalog/gp_warehouse.c]: I have move it to extension

**Review by HuSen8891 — APPROVED** [review]

**Review by my-ship-it — APPROVED** [review]

**Pull request merged**